### PR TITLE
adding in-place exp, sigmoid, softplus, sqrt, tanh

### PR DESCRIPTION
**Pull request by soumith**
cunn equivalents are ready as well.

### CODE REVIEW

**Comment by soumith** [comment]
need to update docs.

**Comment by fmassa** [comment]
Those are nice additions to nn!
I just have a small remark regarding some specific combinations of inplace modules in the backward pass. If a module uses its output to compute the gradient, then inplace operation in the next model shouldn't be performed. Also, two inplace modules can't be used in sequence.
For example, the following network
```lua
m = nn.Sequential():add(nn.Exp()):add(nn.Sqrt(true))
```
won't compute correct gradients in the `Exp` module, because the backward pass in `Exp` uses `self.output`, which was changed in the inplace operation of `Sqrt`. It doesn't happen offen, but it might be difficult to catch it.

As most of `nn` modules, including convolution, pooling and linear, doesn't use `self.output` to compute their gradients, most networks will work just fine with those inplace modules though.

**Comment by nicholas-leonard** [comment]
All those exceptions should be included in the doc.

Nicholas Léonard
450-210-1214

**Comment by AjayTalati** [comment]
The caveats @fmassa has detailed should as be added to the `nngraph` docs. It's common to place nodes within nodes, especially exponentials and square roots.

**Comment by fmassa** [comment]
For bijective functions, in-place operations can be concatenated together, but for that the backward pass should restore the input's module to its initial value. This introduces a few more computations in the backward pass, but maybe it's worth it ?

File: Exp.lua [hunk 1]
```diff
@@ -1,9 +1,30 @@
-local Exp = torch.class('nn.Exp', 'nn.Module')
+local Exp, parent = torch.class('nn.Exp', 'nn.Module')
+
+function Exp:__init(ip)
+   parent.__init(self)
+   -- default for inplace is false
+   self.inplace = ip or false
+   if (ip and type(ip) ~= 'boolean') then
```
Comment (dominikgrewe) [inline Exp.lua]: Nit: No need for parentheses around conditions in Lua.
Same elsewhere.

**Comment by dominikgrewe** [comment]
The code looks good to me, but I agree that the possible pitfalls should be clearly documented, otherwise it's easy to shoot yourself in the foot.

> For bijective functions, in-place operations can be concatenated together, but for that the backward pass should restore the input's module to its initial value. This introduces a few more computations in the backward pass, but maybe it's worth it ?

That would require a module to know that the next module is in-place, doesn't it? Which isn't how nn works. Or do you want to re-compute regardless? Some modules seem to do that, e.g. `nn.AddConstant`.

**Comment by fmassa** [comment]
@dominikgrewe given that `nn` modules are blind wrt their neighbor modules, the solution I mentionned is to always restore the input in the backward pass. As the majority of in place modules are pointwise operations, that wouldn't bring much overhead in the end, I think. That's how I did in `AddConstant` and `MulConstant`, as a proof of concept. But if one have several in place modules like that in a network, that could start to become a bit costly for no gains in most of the time.
But maybe is simply better to mention in the docs that one can't use two in place modules one after another, or that some combinations of modules with in place are not supported.

**Comment by andreaskoepf** [comment]
I like the idea of @fmassa to use the inverse to restore inputs during the backward pass. Just as an speed-over-memory alternative: Inplace mode could be enabled in 'evaluation mode' only based on the module's 'train' member (currently used for dropouts and batch-normalization). Of course this would not help if the main motivation to introduce inplace-operations was to reduce the memory-usage during training...
